cpu/stm32f3/periph/i2c.c

```diff
@@ -0,0 +1,490 @@
+/*
+ * Copyright (C) 2014 FU Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup  driver_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation
+ *
+ * @note This implementation only implements the 7-bit addressing mode.
+ *
+ * @author      Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @auhtor      Thomas Eichinge <thomas.eichinger@fu-berlin.de>
+ * @author      Jan Pohlmann <jan-pohlmann@gmx.de>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "cpu.h"
+#include "irq.h"
+#include "mutex.h"
+#include "periph_conf.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/* guard file in case no I2C device is defined */
+#if I2C_NUMOF
+
+/* static function definitions */
+static void _i2c_init(I2C_TypeDef *i2c, int presc, int scll, int sclh, int sdadel, int scldel);
+static void _pin_config(GPIO_TypeDef *port_scl, GPIO_TypeDef *port_sda, int pin_scl, int pin_sda);
+static void _start(I2C_TypeDef *dev, uint8_t address, uint8_t length, uint8_t rw_flag);
+static inline void _read(I2C_TypeDef *dev, char *data, int length);
+static inline void _write(I2C_TypeDef *dev, char *data, int length);
+static inline void _stop(I2C_TypeDef *dev);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[] =  {
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+#if I2C_1_EN
+    [I2C_1] = MUTEX_INIT,
+#endif
+#if I2C_2_EN
+    [I2C_2] = MUTEX_INIT
+#endif
+#if I2C_3_EN
+    [I2C_3] = MUTEX_INIT
+#endif
+};
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    I2C_TypeDef *i2c;
+    GPIO_TypeDef *port_scl;
+    GPIO_TypeDef *port_sda;
+    int pin_scl = 0, pin_sda = 0;
+    int presc, scll, sclh, sdadel, scldel;
+
+    /* read speed configuration (for i2cclk = 8mhz) */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            presc = 1;
+            scll = 0xC7;
+            sclh = 0xC3;
+            sdadel = 0x2;
+            scldel = 0x4;
+            break;
+
+        case I2C_SPEED_FAST:
+            presc = 0;
+            scll = 0x9;
+            sclh = 0x3;
+            sdadel = 0x1;
+            scldel = 0x3;
+            break;
+
+        case I2C_SPEED_FAST_PLUS:
+            presc = 0;
+            scll = 0x6;
+            sclh = 0x3;
+            sdadel = 0x1;
+            scldel = 0x1;
+            break;
```

Comment:
Haven't found any useful defines in this header, should i create some defines in a *i2ctiming_cpu.h* file in *stm32f3/include*?